# Entregable 3

In [27]:
import numpy as np
import pandas as pd
import io

from bs4 import BeautifulSoup as bs
import requests

## 🎈 1. Ejercicio: Arkansas 🎈


Queremos obtener de la página de Wikipedia de "Arkansas", https://es.wikipedia.org/wiki/Arkansas, la tabla que nos indica la composición étnica a aprtir del censo de EEUU de 2020.

- Revisar la página web de Wikipedia y obtener la tabla en un df
- Sobre el df: revisar qué tipo de columnas tenemos y que valores contienen las mismas
- Renombrar las columnas a nombres más descriptivos
- Eliminar columnas que no sean necesarias
- Cambiar a tipo numérico (realizando los cambios necesarios, %) las columnas que contienen números
- **Realizar un análisis exploratorio sobre la tabla**

In [4]:
url = 'https://es.wikipedia.org/wiki/Arkansas'

In [5]:
tablas = pd.read_html(url)

for i, t in enumerate(tablas):
    print(f"Tabla {i}:\n", t.head(), "\n")

HTTPError: HTTP Error 403: Forbidden

In [6]:
# URL de Wikipedia
url = 'https://es.wikipedia.org/wiki/Arkansas'

# Simulamos un navegador para evitar el error 403
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
                  'AppleWebKit/537.36 (KHTML, like Gecko) '
                  'Chrome/123.0 Safari/537.36'
}

# Realizar la solicitud GET con el User-Agent especificado
response = requests.get(url, headers=headers)

In [7]:
print(f"response: {response}")
print(f"response.status_code: {response.status_code}")

response: <Response [200]>
response.status_code: 200


In [8]:
type(response.text)

str

In [9]:
# Verificar si la solicitud fue exitosa
if response.status_code == 200:
    # Utilizar pandas.read_html para extraer la tabla de la respuesta HTML
    tables = pd.read_html(response.text)
    print(f"El número de tablas encontradas: {len(tables)} \n\n")
else:
    print(f"Error: {response.status_code}")

El número de tablas encontradas: 7 




/tmp/ipykernel_22592/2768303041.py:4: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  tables = pd.read_html(response.text)


In [10]:
io.StringIO(response.text)

In [11]:
# URL de Wikipedia
url = 'https://es.wikipedia.org/wiki/Arkansas'

# Simulamos un navegador para evitar el error 403
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
                  'AppleWebKit/537.36 (KHTML, like Gecko) '
                  'Chrome/123.0 Safari/537.36'
}

# Realizar la solicitud GET con el User-Agent especificado
response = requests.get(url, headers=headers)

if response.status_code == 200:
    # Convertir el HTML en un archivo de html
    html_content = io.StringIO(response.text)

    # Extraer tablas sin warning
    tables = pd.read_html(html_content)

    print(f"El número de tablas encontradas: {len(tables)} \n")
else:
    print(f"Error: {response.status_code}")

El número de tablas encontradas: 7 



In [12]:
type(tables)

list

In [13]:
# Mostramos la primera tabla
display(tables[0].head())

,Arkansas State of Arkansas,Arkansas State of Arkansas.1,Arkansas State of Arkansas.2
0,Estado federado,Estado federado,Estado federado
1,Bandera Escudo,Bandera Escudo,Bandera Escudo
2,Bandera,Escudo,NaN
3,"Otros nombres: The Natural State, Land of Oppo...","Otros nombres: The Natural State, Land of Oppo...","Otros nombres: The Natural State, Land of Oppo..."
4,Lema: En latín: Regnat populus (En español: E...,Lema: En latín: Regnat populus (En español: E...,Lema: En latín: Regnat populus (En español: E...


Como hay 7 tablas, vamos a proceder a buscar la tabla que contiene la información de la composición étnica que nos pide el ejercicio. Echamos un vistazo previo a la página web solicitada para saber qué terminos podemos aplicar para encontrar la tabla que necesitamos. En este caso, sabemos que debe contener el término 'Raza'

In [14]:
url = 'https://es.wikipedia.org/wiki/Juegos_Ol%C3%ADmpicos'

# Simulamos un navegador para evitar el error 403
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
                  'AppleWebKit/537.36 (KHTML, like Gecko) '
                  'Chrome/123.0 Safari/537.36'
}

# Realizar la solicitud GET con el User-Agent especificado
response = requests.get(url, headers=headers)

if response.status_code == 200:
    # Convertir el HTML en un archivo de html
    html_content = io.StringIO(response.text)

    # Extraer tablas sin warning
    # # Filtramos las tablas que contienen el texto "Raza" en el encabezado
    tables = pd.read_html(html_content, match='Raza')

    print(f"El número de tablas encontradas: {len(tables)} \n")
else:
    print(f"Error: {response.status_code}")

ValueError: No tables found matching pattern 'Raza'

In [15]:
#Como nos da error con varias palabras, haremos la búsqueda de la tabla de forma manual

display(tables[4].head())

,Raza y etnia[11]​,Alone,Alone.1,Total,Total.1
0,Blanco no hispano,68.5%,NaN,73.2%,NaN
1,Negros,14.9%,NaN,16.2%,NaN
2,Hispanios o Latinos[12]​,—,NaN,8.5%,NaN
3,Asiáticos,1.7%,NaN,2.2%,NaN
4,Nativos,0.7%,NaN,3.4%,NaN


In [16]:
#Convertimos a dataframe para poder trabajar con los datos

df = tables[4]
df

,Raza y etnia[11]​,Alone,Alone.1,Total,Total.1
0,Blanco no hispano,68.5%,NaN,73.2%,NaN
1,Negros,14.9%,NaN,16.2%,NaN
2,Hispanios o Latinos[12]​,—,NaN,8.5%,NaN
3,Asiáticos,1.7%,NaN,2.2%,NaN
4,Nativos,0.7%,NaN,3.4%,NaN
5,Estadounidenses de las islas del Pacífico,0.5%,NaN,0.6%,NaN
6,Otro,0.3%,NaN,1.1%,NaN


In [17]:
#Comprobamos el tipo de datos que tenemos en el df

df.dtypes

,0
Raza y etnia[11]​,object
Alone,object
Alone.1,float64
Total,object
Total.1,float64


In [19]:
#Eliminamos las columnas sin información

df_1 = df.drop(columns=['Alone.1', 'Total.1'])
df_1

,Raza y etnia[11]​,Alone,Total
0,Blanco no hispano,68.5%,73.2%
1,Negros,14.9%,16.2%
2,Hispanios o Latinos[12]​,—,8.5%
3,Asiáticos,1.7%,2.2%
4,Nativos,0.7%,3.4%
5,Estadounidenses de las islas del Pacífico,0.5%,0.6%
6,Otro,0.3%,1.1%


In [20]:
#Renombramos las columnas para que sean más claras

df_1 = df_1.rename(columns={
    'Raza y etnia[11]​': 'etnia',
    'Alone': 'porcentaje'
})

In [21]:
df_1

,etnia,porcentaje,Total
0,Blanco no hispano,68.5%,73.2%
1,Negros,14.9%,16.2%
2,Hispanios o Latinos[12]​,—,8.5%
3,Asiáticos,1.7%,2.2%
4,Nativos,0.7%,3.4%
5,Estadounidenses de las islas del Pacífico,0.5%,0.6%
6,Otro,0.3%,1.1%


In [22]:
#Realizamos un análisis exploratorio

df_1.describe()

,etnia,porcentaje,Total
count,7,7,7
unique,7,7,7
top,Blanco no hispano,68.5%,73.2%
freq,1,1,1


In [23]:
#Como el anterior código nos ha arrojado resultados inesperados, volvemos a comprobar los datos

df_1.dtypes

,0
etnia,object
porcentaje,object
Total,object


In [24]:
#Transformamos a float

df_1['porcentaje'] = (
    df_1['porcentaje']
    .str.replace('%','', regex=False)
    .str.replace(',','.', regex=False)
    .str.replace('—','', regex=False)
    .str.strip()
    .replace('', None)
    .astype(float)
)

df_1['Total'] = (
    df_1['Total']
    .str.replace('%','', regex=False)
    .str.replace(',','.', regex=False)
    .str.replace('—','', regex=False)
    .str.strip()
    .replace('', None)
    .astype(float)
)

In [25]:
df_1.dtypes

,0
etnia,object
porcentaje,float64
Total,float64


In [26]:
df_1[['porcentaje','Total']].describe()

,porcentaje,Total
count,6.000000,7.000000
mean,14.433333,15.028571
std,27.085248,26.235772
min,0.300000,0.600000
25%,0.550000,1.650000
50%,1.200000,3.400000
75%,11.600000,12.350000
max,68.500000,73.200000


La tabla inicial recogida de Wikipedia ya nos da los datos ordenados que nos muestran que la etnia dominante es la 'Blanco no hispano', con un 73%, seguida de los 'Negros' con un 16%.

## 🎈 2. Ejercicio pandas 🎈


- Pandas: scrapear la tabla que contiene informacion sobre los álbums más vendidos por año en todo el mundo (https://es.wikipedia.org/wiki/Anexo:%C3%81lbumes_musicales_m%C3%A1s_vendidos) generando un dataframe con la información del año, nombre del álbum, artista y ventas. Ésta ultima columna deberá tener sus valores en numeros enteros en millones.

In [40]:
url = 'https://es.wikipedia.org/wiki/Anexo:%C3%81lbumes_musicales_m%C3%A1s_vendidos)'

In [41]:
tablas = pd.read_html(url)

for i, t in enumerate(tablas):
    print(f"Tabla {i}:\n", t.head(), "\n")

HTTPError: HTTP Error 403: Forbidden

In [46]:
# URL de Wikipedia
url = 'https://es.wikipedia.org/wiki/Anexo:%C3%81lbumes_musicales_m%C3%A1s_vendidos)'
# Simulamos un navegador para evitar el error 403
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
                  'AppleWebKit/537.36 (KHTML, like Gecko) '
                  'Chrome/123.0 Safari/537.36'
}

# Realizar la solicitud GET con el User-Agent especificado
response = requests.get(url, headers=headers)

In [47]:
print(f"response: {response}")
print(f"response.status_code: {response.status_code}")

response: <Response [404]>
response.status_code: 404


Como nos da el código 404, vamos a probar a cambiar la url por la actual

In [49]:
# URL de Wikipedia
url = 'https://es.wikipedia.org/wiki/Anexo:%C3%81lbumes_musicales_m%C3%A1s_vendidos#%C3%81lbum_m%C3%A1s_vendido_por_a%C3%B1o_en_todo_el_mundo'

# Simulamos un navegador para evitar el error 403
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
                  'AppleWebKit/537.36 (KHTML, like Gecko) '
                  'Chrome/123.0 Safari/537.36'
}

# Realizar la solicitud GET con el User-Agent especificado
response = requests.get(url, headers=headers)

In [50]:
print(f"response: {response}")
print(f"response.status_code: {response.status_code}")

response: <Response [200]>
response.status_code: 200


In [51]:
type(response.text)

str

In [53]:
# Verificar si la solicitud fue exitosa
if response.status_code == 200:
    # Utilizar pandas.read_html para extraer la tabla de la respuesta HTML
    tables = pd.read_html(response.text)
    print(f"El número de tablas encontradas: {len(tables)} \n\n")
else:
    print(f"Error: {response.status_code}")

El número de tablas encontradas: 7 




/tmp/ipykernel_22592/2768303041.py:4: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  tables = pd.read_html(response.text)


In [52]:
io.StringIO(response.text)

In [54]:
# URL de Wikipedia
url = 'https://es.wikipedia.org/wiki/Anexo:%C3%81lbumes_musicales_m%C3%A1s_vendidos#%C3%81lbum_m%C3%A1s_vendido_por_a%C3%B1o_en_todo_el_mundo'

# Simulamos un navegador para evitar el error 403
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
                  'AppleWebKit/537.36 (KHTML, like Gecko) '
                  'Chrome/123.0 Safari/537.36'
}

# Realizar la solicitud GET con el User-Agent especificado
response = requests.get(url, headers=headers)

if response.status_code == 200:
    # Convertir el HTML en un archivo de html
    html_content = io.StringIO(response.text)

    # Extraer tablas sin warning
    tables = pd.read_html(html_content)

    print(f"El número de tablas encontradas: {len(tables)} \n")
else:
    print(f"Error: {response.status_code}")


El número de tablas encontradas: 7 



In [55]:
type(tables)

list

In [56]:
# Mostramos la primera tabla
display(tables[0].head())

,0
0,Álbum de estudio
1,Álbum recopilatorio
2,Banda sonora
3,Álbum en vivo


In [60]:
#Buscamos la tabla que necesitamos

url = 'https://es.wikipedia.org/wiki/Anexo:%C3%81lbumes_musicales_m%C3%A1s_vendidos#%C3%81lbum_m%C3%A1s_vendido_por_a%C3%B1o_en_todo_el_mundo'

# Simulamos un navegador para evitar el error 403
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
                  'AppleWebKit/537.36 (KHTML, like Gecko) '
                  'Chrome/123.0 Safari/537.36'
}

# Realizar la solicitud GET con el User-Agent especificado
response = requests.get(url, headers=headers)

if response.status_code == 200:
    # Convertir el HTML en un archivo de html
    html_content = io.StringIO(response.text)

    # Extraer tablas sin warning
    # # Filtramos las tablas que contienen el texto "Ref" en el encabezado
    tables = pd.read_html(html_content, match='Ref')

    print(f"El número de tablas encontradas: {len(tables)} \n")
else:
    print(f"Error: {response.status_code}")

El número de tablas encontradas: 1 



In [61]:
#Esta es la tabla que coincide con lo pedido en el ejercicio

display(tables[0].head())

,Año,Álbum,Artista(s),Ventas (millones),Ref(s)
0,2001,Hybrid Theory,Linkin Park,9.0,[191]​
1,2002,The Eminem Show,Eminem,13.9,[192]​
2,2003,Come Away with Me,Norah Jones,11.0,[193]​
3,2004,Confessions,Usher,12.0,[194]​
4,2005,X&Y,Coldplay,8.3,[195]​


In [62]:
#Convertimos a dataframe para poder trabajar con los datos

df_a = tables[0]
df_a

,Año,Álbum,Artista(s),Ventas (millones),Ref(s)
0,2001,Hybrid Theory,Linkin Park,9.00,[191]​
1,2002,The Eminem Show,Eminem,13.90,[192]​
2,2003,Come Away with Me,Norah Jones,11.00,[193]​
3,2004,Confessions,Usher,12.00,[194]​
4,2005,X&Y,Coldplay,8.30,[195]​
5,2006,High School Musical,Artistas varios,7.00,[196]​
6,2007,High School Musical 2,Artistas varios,6.00,[197]​
7,2008,Viva la Vida or Death and All His Friends,Coldplay,6.80,[198]​
8,2009,I Dreamed a Dream,Susan Boyle,8.30,[199]​
9,2010,Recovery,Eminem,5.70,[200]​


In [64]:
#Comprobamos las columnas del df

df_a.columns

Index(['Año', 'Álbum', 'Artista(s)', 'Ventas (millones)', 'Ref(s)'], dtype='object')

In [66]:
#Comprobamos el tipo de datos que tenemos en el df

df_a.dtypes

,0
Año,int64
Álbum,object
Artista(s),object
Ventas (millones),float64
Ref(s),object


In [67]:
#Eliminamos las columnas sin información

df_a = df_a.drop(columns=['Ref(s)'])
df_a

,Año,Álbum,Artista(s),Ventas (millones)
0,2001,Hybrid Theory,Linkin Park,9.00
1,2002,The Eminem Show,Eminem,13.90
2,2003,Come Away with Me,Norah Jones,11.00
3,2004,Confessions,Usher,12.00
4,2005,X&Y,Coldplay,8.30
5,2006,High School Musical,Artistas varios,7.00
6,2007,High School Musical 2,Artistas varios,6.00
7,2008,Viva la Vida or Death and All His Friends,Coldplay,6.80
8,2009,I Dreamed a Dream,Susan Boyle,8.30
9,2010,Recovery,Eminem,5.70


In [69]:
#Cambiamos el tipo de número en la columna de ventas

df_a['Ventas (millones)'] = df_a['Ventas (millones)'].astype(float) * 1_000_000
df_a

,Año,Álbum,Artista(s),Ventas (millones)
0,2001,Hybrid Theory,Linkin Park,9000000.0
1,2002,The Eminem Show,Eminem,13900000.0
2,2003,Come Away with Me,Norah Jones,11000000.0
3,2004,Confessions,Usher,12000000.0
4,2005,X&Y,Coldplay,8300000.0
5,2006,High School Musical,Artistas varios,7000000.0
6,2007,High School Musical 2,Artistas varios,6000000.0
7,2008,Viva la Vida or Death and All His Friends,Coldplay,6800000.0
8,2009,I Dreamed a Dream,Susan Boyle,8300000.0
9,2010,Recovery,Eminem,5700000.0


In [70]:
#Renombramos las columnas para que sean más claras

df_a = df_a.rename(columns={
    'Ventas (millones)​': 'ventas'
})

In [71]:
df_a

,Año,Álbum,Artista(s),Ventas (millones)
0,2001,Hybrid Theory,Linkin Park,9000000.0
1,2002,The Eminem Show,Eminem,13900000.0
2,2003,Come Away with Me,Norah Jones,11000000.0
3,2004,Confessions,Usher,12000000.0
4,2005,X&Y,Coldplay,8300000.0
5,2006,High School Musical,Artistas varios,7000000.0
6,2007,High School Musical 2,Artistas varios,6000000.0
7,2008,Viva la Vida or Death and All His Friends,Coldplay,6800000.0
8,2009,I Dreamed a Dream,Susan Boyle,8300000.0
9,2010,Recovery,Eminem,5700000.0


## 🎈 3. Ejercicio beautiful soup🎈


- Beautiful Soup: scrapear la tabla de lenguajes de programación más utilizados (https://www.tiobe.com/tiobe-index/) creando un dataframe con una columna llamada "lenguaje" y otra columna llamada "porcentaje_uso" (cogiendo la info de programming language y ratings respectivamente).

In [114]:
from bs4 import BeautifulSoup as bs
import requests
import pandas as pd

In [115]:
# URL
url = 'https://www.tiobe.com/tiobe-index/'

# Simulamos un navegador para evitar el error 403
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
                  'AppleWebKit/537.36 (KHTML, like Gecko) '
                  'Chrome/123.0 Safari/537.36'
}

# Realizar la solicitud GET con el User-Agent especificado
response = requests.get(url, headers=headers)

if response.status_code == 200:
    print(f"status_code: {response.status_code}")
    # Convertir el HTML en un archivo de html
    soup = bs(response.text, 'html.parser')
else:
    print(f"Error: {response.status_code}")


status_code: 200


In [116]:
# soup es un tipo de dato propio de bs
# representa un HTML parseado
type(soup)

bs4.BeautifulSoup

In [117]:
soup.prettify()

'<!DOCTYPE html>\n<html lang="en-US">\n <head>\n  <meta charset="utf-8"/>\n  <meta content="width=device-width, initial-scale=1.0" name="viewport"/>\n  <title>\n   TIOBE Index - TIOBE\n  </title>\n  <link href="http://gmpg.org/xfn/11" rel="profile"/>\n  <link href="https://www.tiobe.com/wp-content/themes/tiobe/style.css" media="screen" rel="stylesheet" type="text/css"/>\n  <link href="https://www.tiobe.com/xmlrpc.php" rel="pingback"/>\n  <link href="https://www.tiobe.com/wp-content/themes/tiobe/favicon/apple-touch-icon.png" rel="apple-touch-icon" sizes="180x180"/>\n  <link href="https://www.tiobe.com/wp-content/themes/tiobe/favicon/favicon-32x32.png" rel="icon" sizes="32x32" type="image/png"/>\n  <link href="https://www.tiobe.com/wp-content/themes/tiobe/favicon/favicon-16x16.png" rel="icon" sizes="16x16" type="image/png"/>\n  <link href="https://www.tiobe.com/wp-content/themes/tiobe/favicon/site.webmanifest" rel="manifest"/>\n  <link color="#287ac8" href="https://www.tiobe.com/wp-conte

In [118]:
soup


<!DOCTYPE html>

<html lang="en-US">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<title>TIOBE Index - TIOBE</title>
<link href="http://gmpg.org/xfn/11" rel="profile"/>
<link href="https://www.tiobe.com/wp-content/themes/tiobe/style.css" media="screen" rel="stylesheet" type="text/css"/>
<link href="https://www.tiobe.com/xmlrpc.php" rel="pingback"/>
<link href="https://www.tiobe.com/wp-content/themes/tiobe/favicon/apple-touch-icon.png" rel="apple-touch-icon" sizes="180x180"/>
<link href="https://www.tiobe.com/wp-content/themes/tiobe/favicon/favicon-32x32.png" rel="icon" sizes="32x32" type="image/png"/>
<link href="https://www.tiobe.com/wp-content/themes/tiobe/favicon/favicon-16x16.png" rel="icon" sizes="16x16" type="image/png"/>
<link href="https://www.tiobe.com/wp-content/themes/tiobe/favicon/site.webmanifest" rel="manifest"/>
<link color="#287ac8" href="https://www.tiobe.com/wp-content/themes/tiobe/favicon/safari-pinned-tab.sv

In [119]:
tables = soup.find_all('table')

In [120]:
type(tables)

bs4.element.ResultSet

In [121]:
len(tables)

4

In [122]:
# Buscar la tabla específica basada en el caption
# Utilizamos soup.find_all para encontrar todas las tablas (<table>)
#  que tienen la clase 'table-top20' en el documento HTML
# filtrando por la clase se reduce el numero de coincidencias
tables = soup.find_all('table', {'class': 'table-top20'})

print(f"Hemos encontrado {len(tables)} tablas que tienen en la clase 'table-top20' en nuestro documento HTML ")

print(f"\nSi no filtráramos por la clase, encontraríamos {len(soup.find_all('table'))} tablas en nuestro documento HTML ")

Hemos encontrado 1 tablas que tienen en la clase 'table-top20' en nuestro documento HTML 

Si no filtráramos por la clase, encontraríamos 4 tablas en nuestro documento HTML 


In [123]:
import pandas as pd

df_p = pd.read_html(str(tables[0]))[0]
df_p

/tmp/ipykernel_22592/151933729.py:3: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  df_p = pd.read_html(str(tables[0]))[0]


,Mar 2026,Mar 2025,Change,Programming Language,Programming Language.1,Ratings,Change.1
0,1,1,NaN,NaN,Python,21.25%,-2.59%
1,2,4,NaN,NaN,C,11.55%,+2.02%
2,3,2,NaN,NaN,C++,8.18%,-2.90%
3,4,3,NaN,NaN,Java,7.99%,-2.37%
4,5,5,NaN,NaN,C#,6.36%,+1.49%
5,6,6,NaN,NaN,JavaScript,3.45%,-0.01%
6,7,9,NaN,NaN,Visual Basic,2.50%,-0.02%
7,8,8,NaN,NaN,SQL,2.00%,-0.57%
8,9,16,NaN,NaN,R,1.88%,+0.94%
9,10,10,NaN,NaN,Delphi/Object Pascal,1.80%,-0.36%


In [124]:
df_p.head()
df_p.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Mar 2026                20 non-null     int64  
 1   Mar 2025                20 non-null     int64  
 2   Change                  0 non-null      float64
 3   Programming Language    0 non-null      float64
 4   Programming Language.1  20 non-null     object 
 5   Ratings                 20 non-null     object 
 6   Change.1                20 non-null     object 
dtypes: float64(2), int64(2), object(3)
memory usage: 1.2+ KB


In [125]:
df_p.columns

Index(['Mar 2026', 'Mar 2025', 'Change', 'Programming Language',
       'Programming Language.1', 'Ratings', 'Change.1'],
      dtype='object')

In [126]:
df_p = df_p.drop(columns=['Programming Language', 'Change'])
df_p

,Mar 2026,Mar 2025,Programming Language.1,Ratings,Change.1
0,1,1,Python,21.25%,-2.59%
1,2,4,C,11.55%,+2.02%
2,3,2,C++,8.18%,-2.90%
3,4,3,Java,7.99%,-2.37%
4,5,5,C#,6.36%,+1.49%
5,6,6,JavaScript,3.45%,-0.01%
6,7,9,Visual Basic,2.50%,-0.02%
7,8,8,SQL,2.00%,-0.57%
8,9,16,R,1.88%,+0.94%
9,10,10,Delphi/Object Pascal,1.80%,-0.36%


In [127]:
df_p = df_p.rename(columns={
    'Programming Language.1​': 'lenguaje',
    'Ratings': 'porcentaje_uso'
})

In [128]:
df_p

,Mar 2026,Mar 2025,Programming Language.1,porcentaje_uso,Change.1
0,1,1,Python,21.25%,-2.59%
1,2,4,C,11.55%,+2.02%
2,3,2,C++,8.18%,-2.90%
3,4,3,Java,7.99%,-2.37%
4,5,5,C#,6.36%,+1.49%
5,6,6,JavaScript,3.45%,-0.01%
6,7,9,Visual Basic,2.50%,-0.02%
7,8,8,SQL,2.00%,-0.57%
8,9,16,R,1.88%,+0.94%
9,10,10,Delphi/Object Pascal,1.80%,-0.36%


In [129]:
print(df_p.columns.tolist())


['Mar 2026', 'Mar 2025', 'Programming Language.1', 'porcentaje_uso', 'Change.1']


In [130]:
df_p.columns = (
    df_p.columns
    .str.strip()
    .str.replace('\u200b','')
)

In [131]:
df_p = df_p.rename(columns={
    'Programming Language.1​': 'lenguaje'
})

In [132]:
df_p

,Mar 2026,Mar 2025,Programming Language.1,porcentaje_uso,Change.1
0,1,1,Python,21.25%,-2.59%
1,2,4,C,11.55%,+2.02%
2,3,2,C++,8.18%,-2.90%
3,4,3,Java,7.99%,-2.37%
4,5,5,C#,6.36%,+1.49%
5,6,6,JavaScript,3.45%,-0.01%
6,7,9,Visual Basic,2.50%,-0.02%
7,8,8,SQL,2.00%,-0.57%
8,9,16,R,1.88%,+0.94%
9,10,10,Delphi/Object Pascal,1.80%,-0.36%


In [133]:
col_old = [c for c in df_p.columns if 'Programming Language' in c][0]
df_p = df_p.rename(columns={col_old: 'lenguaje'})

In [134]:
df_p

,Mar 2026,Mar 2025,lenguaje,porcentaje_uso,Change.1
0,1,1,Python,21.25%,-2.59%
1,2,4,C,11.55%,+2.02%
2,3,2,C++,8.18%,-2.90%
3,4,3,Java,7.99%,-2.37%
4,5,5,C#,6.36%,+1.49%
5,6,6,JavaScript,3.45%,-0.01%
6,7,9,Visual Basic,2.50%,-0.02%
7,8,8,SQL,2.00%,-0.57%
8,9,16,R,1.88%,+0.94%
9,10,10,Delphi/Object Pascal,1.80%,-0.36%


In [135]:
#Eliminamos las columnas para obtener df final

df_p = df_p.drop(columns=['Mar 2026', 'Mar 2025', 'Change.1'])
df_p

,lenguaje,porcentaje_uso
0,Python,21.25%
1,C,11.55%
2,C++,8.18%
3,Java,7.99%
4,C#,6.36%
5,JavaScript,3.45%
6,Visual Basic,2.50%
7,SQL,2.00%
8,R,1.88%
9,Delphi/Object Pascal,1.80%
